equational_theories/Closure.lean

```diff
@@ -0,0 +1,244 @@
+
+def Outcome.isTrue : Outcome → Bool
+  | explicit_proof_true => true
+  | implicit_proof_true => true
+  | explicit_conjecture_true => true
+  | implicit_conjecture_true => true
+  | unknown => false
+  | implicit_conjecture_false => false
+  | explicit_conjecture_false => false
+  | implicit_proof_false => false
+  | explicit_proof_false => false
+
+partial def dfs1 (graph : Array (Array Nat)) (vertex : Nat) (vis : Array Bool) (order : Array Nat) :
+    Array Bool × Array Nat := Id.run do
+  let mut vis1 := vis.set! vertex true
+  let mut ord := order
+  for v in graph[vertex]! do
+    unless vis1[v]! do
+      (vis1, ord) := dfs1 graph v vis1 ord
+  pure (vis1, ord.push vertex)
+
+partial def dfs2 (graph : Array (Array Nat)) (vertex : Nat) (component : Array Nat) (component_id : Nat) :
+    Array Nat := Id.run do
+  let mut comp := component.set! vertex component_id
+  for v in graph[vertex]! do
+    if component[v]! == 0 then
+      comp := dfs2 graph v comp component_id
+  pure comp
+
+def Bitset := Array UInt64
+
+def Bitset.toArray : Bitset → Array UInt64 := id
+
+instance : Inhabited Bitset := ⟨#[]⟩
+
+def Bitset.mk (n : Nat) : Bitset := Array.mkArray ((n + 63) >>> 6) 0
+
+def Bitset.set (b : Bitset) (n : Nat) : Bitset :=
+  b.modify (n >>> 6) (fun x ↦ x ||| (1 <<< ((UInt64.ofNat n) &&& 63)))
+
+def Bitset.get (b : Bitset) (n : Nat) : Bool :=
+  (b.toArray[n >>> 6]! >>> ((UInt64.ofNat n) &&& 63)) &&& 1 != 0
+
+def closure (inp : Array Edge) : Array Edge := Id.run do
+  -- number the equations (arbitrarily) for easier processing
+  let mut eqs : Std.HashMap String Nat := {}
+  let mut eqs_order : Array String := #[]
+  for imp in inp do
+    match eqs.containsThenInsertIfNew imp.lhs eqs.size with
+    | (n, neqs) =>
+      eqs := neqs
+      unless n do eqs_order := eqs_order.push imp.lhs
+    match eqs.containsThenInsertIfNew imp.rhs eqs.size with
+    | (n, neqs) =>
+      eqs := neqs
+      unless n do eqs_order := eqs_order.push imp.rhs
+
+  -- construct the implication/non-implication graph
+  let n := eqs.size
+  let mut graph : Array (Array Nat) := Array.mkArray (2 * n) #[]
+  let mut revgraph : Array (Array Nat) := Array.mkArray (2 * n) #[]
+  for imp in inp do
+    match imp with
+    | .implication imp =>
+      graph := graph.modify (eqs[imp.rhs]!) (fun x => x.push eqs[imp.lhs]!)
+      graph := graph.modify (eqs[imp.rhs]! + n) (fun x => x.push (eqs[imp.lhs]! + n))
+      revgraph := revgraph.modify (eqs[imp.lhs]!) (fun x => x.push eqs[imp.rhs]!)
+      revgraph := revgraph.modify (eqs[imp.lhs]! + n) (fun x => x.push (eqs[imp.rhs]! + n))
+    | .nonimplication imp =>
+      graph := graph.modify (eqs[imp.lhs]!) (fun x => x.push (eqs[imp.rhs]! + n))
+      revgraph := revgraph.modify (eqs[imp.rhs]! + n) (fun x => x.push eqs[imp.lhs]!)
+
+  let mut vis : Array Bool := Array.mkArray (2 * n) false
+  let mut order : Array Nat := Array.mkEmpty (2 * n)
+
+  -- compute SCCs and the condensation graph using Kosaraju's algorithm
```

Comment:
Please expand SCC. "stongly connected components", I gather?

Reply:
Yes, I'll expand it